lib/acts_as_taggable_on_mongoid/taggable/core.rb

```diff
@@ -40,14 +40,98 @@ def taggable_mixin
         end
       end
 
-      def apply_pre_processed_defaults
-        super
+      def apply_post_processed_defaults
+        defaults = super
 
-        set_tag_list_defaults
+        defaults |= set_tag_list_defaults
+
+        defaults
       end
 
       private
 
+      def set_tag_list(tag_definition, new_tags)
```

Comment:
A bunch of utility functions that do the work of the `<tag_list>_function` methods by passing in the `tag_definition` to drive the work.

The code was moved here to simplify the code a lot and to pass Rubocop.